File: dynamic-map/src/types/cognitiveHub.ts
export type CognitiveLevel = 1 | 2 | 3 | 4 | 5

export type DurationMinutes = 17 | 35 | 50

export type RateOfChange = 'Low' | 'Medium' | 'High'

export type Season =
  | 'Battle'
  | 'Flow'
  | 'Exhaustion/Recovery'
  | 'Exhaustion v2'
  | 'Frustration Season'
  | 'Doubt'
  | 'Grounding'

export const ENTRY_SEASONS: Season[] = ['Battle', 'Flow', 'Exhaustion/Recovery']

export const DEVIATION_SEASONS: Season[] = [
  'Exhaustion v2',
  'Frustration Season',
  'Doubt',
  'Grounding',
]

export type PitstopKind =
  | 'start'
  | 'work'
  | 'peak'
  | 'recovery'
  | 'grounding'
  | 'stranding'
  | 'end'

export interface MapCoordinates {
  x: number
  y: number
}

export interface Pitstop {
  id: string
  label: string
  level: CognitiveLevel
  season: Season
  kind: PitstopKind
  /** Offset in ms from session start at which this pitstop is scheduled. */
  scheduledTime: number
  /** Epoch ms at which this pitstop was actually created/triggered (set for stranding markers, and for the start pitstop). */
  realTimeTrigger?: number
  mapCoordinates: MapCoordinates
}

export interface PacingPath {
  pitstops: Pitstop[]
  totalDurationMs: number
}

export interface Task {
  id: string
  name: string
  durationMinutes: DurationMinutes
  /** Minimum cognition needed to make progress on the task — the floor of the pacing range. */
  minCognitiveLevel: CognitiveLevel
  /** Upper cognitive load limit — the ceiling of the pacing range. */
  cognitiveLimit: CognitiveLevel
  entrySeason: Season
  rateOfChange: RateOfChange
  createdAt: number
  /** Optional sticky-note prompt to keep front-of-mind for the session — visible on every device. */
  guidingQuestion?: string
}

export interface DeviationEvent {
  id: string
  timestamp: number
  elapsedAtTrigger: number
  season: Season
  pitstopId: string
}

export type PitstopStatus = 'past' | 'active' | 'next' | 'future' | 'deviation'

export interface SessionState {
  task: Task
  path: PacingPath
  deviations: DeviationEvent[]
  elapsedTime: number
  startTime: number | null
  isActive: boolean
  isPaused: boolean
}

/**
 * Whether the cognitive-level loop oscillates up and back down within a cycle
 * ('mirror', e.g. 1→2→3→2→1→2→3…) or sweeps up and resets to the floor each
 * cycle ('ramp', e.g. 1→2→3→1→2→3…).
 */
export type CyclePattern = 'mirror' | 'ramp'

/**
 * How long each pacing segment lasts relative to the others: 'even' keeps
 * segment durations roughly equal, 'negative' front-loads longer segments
 * that shrink over the session, 'positive' starts short and lengthens.
 */
export type SplitType = 'even' | 'negative' | 'positive'

export interface UserPreferences {
  cyclePattern: CyclePattern
  splitType: SplitType
  /** Default entry season pre-filled on the task form — the emotional/energy state the user most often starts from. */
  preferredSeason: Season
  updatedAt: number
}

File: watch-app/CognitivePacingAtlasWatch/Models/CognitiveHub.swift
import Foundation

// Swift port of dynamic-map/src/types/cognitiveHub.ts.
// Field names match the JS-written Firestore documents exactly so Codable
// decoding works with no custom CodingKeys.

/// 1...5, clamped via PacingAlgorithm.clampLevel. Kept as a plain Int (rather
/// than an enum) to mirror the TS `1 | 2 | 3 | 4 | 5` numeric-literal union.
typealias CognitiveLevel = Int

enum RateOfChange: String, Codable {
    case low = "Low"
    case medium = "Medium"
    case high = "High"
}

enum Season: String, Codable {
    case battle = "Battle"
    case flow = "Flow"
    case exhaustionRecovery = "Exhaustion/Recovery"
    case exhaustionV2 = "Exhaustion v2"
    case frustrationSeason = "Frustration Season"
    case doubt = "Doubt"
    case grounding = "Grounding"
}

enum PitstopKind: String, Codable {
    case start, work, peak, recovery, grounding, stranding, end
}

/// 'mirror': oscillates up and back down within a cycle (2 -> 3 -> 4 -> 3 -> 2 -> …).
/// 'ramp': sweeps floor to ceiling and resets each cycle (1 -> 2 -> 3 -> 1 -> 2 -> 3 -> …).
enum CyclePattern: String, Codable {
    case mirror, ramp
}

/// 'even' keeps segment durations roughly equal, 'negative' front-loads longer
/// segments that shrink over the session, 'positive' starts short and lengthens.
enum SplitType: String, Codable {
    case even, negative, positive
}

enum PitstopStatus {
    case past, active, next, future, deviation
}

struct MapCoordinates: Codable, Equatable {
    var x: Double
    var y: Double
}

struct Pitstop: Codable, Identifiable, Equatable {
    var id: String
    var label: String
    var level: CognitiveLevel
    var season: Season
    var kind: PitstopKind
    /// Offset in ms from session start at which this pitstop is scheduled.
    var scheduledTime: Double
    /// Epoch ms at which this pitstop was actually created/triggered.
    var realTimeTrigger: Double?
    var mapCoordinates: MapCoordinates
}

struct PacingPath: Codable, Equatable {
    var pitstops: [Pitstop]
    var totalDurationMs: Double
}

/// Named WatchTask (not Task) to avoid colliding with Swift's concurrency Task type.
struct WatchTask: Codable, Equatable {
    var id: String
    var name: String
    var durationMinutes: Int
    /// Minimum cognition needed to make progress — the floor of the pacing range.
    var minCognitiveLevel: CognitiveLevel
    /// Upper cognitive load limit — the ceiling of the pacing range.
    var cognitiveLimit: CognitiveLevel
    var entrySeason: Season
    var rateOfChange: RateOfChange
    var createdAt: Double
}

struct DeviationEvent: Codable, Identifiable, Equatable {
    var id: String
    var timestamp: Double
    var elapsedAtTrigger: Double
    var season: Season
    var pitstopId: String
}

struct SessionState: Codable, Equatable {
    var task: WatchTask
    var path: PacingPath
    var deviations: [DeviationEvent]
    var elapsedTime: Double
    var startTime: Double?
    var isActive: Bool
    var isPaused: Bool
}

struct UserPreferences: Codable, Equatable {
    var cyclePattern: CyclePattern
    var splitType: SplitType
    /// Default entry season — the emotional/energy state the user most often starts from.
    var preferredSeason: Season
    var updatedAt: Double
}
